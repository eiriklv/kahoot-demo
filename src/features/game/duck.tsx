import { RootState } from "../../root-reducer";
import { createSelector } from "reselect";
import { createPartitions } from "./utils";
import { currentDefinitions, startBoard } from "./data";

/**
 * Feature identifier
 */
export const NAME = "@game";

/**
 * Types, interfaces and constants
 */
export const RESET_GAME = "@game/RESET_GAME";
export const COLLECT_ITEM = "@game/COLLECT_ITEM";

export interface ResetGameAction {
  type: typeof RESET_GAME;
}

export interface CollectItemAction {
  type: typeof COLLECT_ITEM;
  payload: {
    index: number;
  };
}

export type GameAction = ResetGameAction | CollectItemAction;

export type BoardItem = "A" | "B" | "C" | "D";

export interface ScoringDefinition {
  type: BoardItem;
  points: number;
  bonus: number[];
}

export interface ScoreBoardItem {
  type: BoardItem;
  quantity: number;
  points: number;
}

export interface GameState {
  board: BoardItem[];
  collected: BoardItem[];
  definitions: ScoringDefinition[];
}

/**
 * Action creators
 */
export const resetGame = (): ResetGameAction => {
  return {
    type: RESET_GAME,
  };
};

export const collectItem = (index: number): CollectItemAction => {
  return {
    type: COLLECT_ITEM,
    payload: { index },
  };
};

/**
 * Selectors
 */
export const getState = (state: RootState) => state[NAME];
export const getGameBoard = (state: RootState) => getState(state).board;
export const getDefinitions = (state: RootState) => getState(state).definitions;
export const getCollectedItems = (state: RootState) =>
  getState(state).collected;

export const getScoreBoard = createSelector(
  getCollectedItems,
  getDefinitions,
  (collectedItems, scoringDefinitions) => {
    const partitionedItems = createPartitions<BoardItem>(collectedItems);
    return partitionedItems.reduce<ScoreBoardItem[]>((output, partition) => {
      const type = partition[0];

      const { points = 0, bonus = [] } =
        scoringDefinitions.find(
          (scoringDefinition) => scoringDefinition.type === type
        ) || {};

      const itemPoints = partition.reduce((sum, item) => sum + points, 0);

      const itemBonus = bonus
        .slice(0, partition.length)
        .reduce((sum, bonus) => sum + bonus, 0);

      return [
        ...output,
        {
          type,
          quantity: partition.length,
          points: itemPoints + itemBonus,
        },
      ];
    }, []);
  }
);

export const getBonusPoints = createSelector(
  getCollectedItems,
  getDefinitions,
  (collectedItems, scoringDefinitions) => {
    const partitionedItems = createPartitions<BoardItem>(collectedItems);
    return partitionedItems.reduce((totalPoints, partition) => {
      const type = partition[0];

      const { bonus = [] } =
        scoringDefinitions.find(
          (scoringDefinition) => scoringDefinition.type === type
        ) || {};

      const bonusPoints = bonus
        .slice(0, partition.length)
        .reduce((sum, bonus) => sum + bonus, 0);

      return totalPoints + bonusPoints;
    }, 0);
  }
);

export const getTotalScore = createSelector(getScoreBoard, (scoreBoard) => {
  return scoreBoard.reduce(
    (total, scoreBoardItem) => scoreBoardItem.points + total,
    0
  );
});

/**
 * Game state and reducer
 */
export const initialGameState: GameState = {
  board: startBoard,
  collected: [],
  definitions: currentDefinitions,
};

export default function reducer(
  state: GameState = initialGameState,
  action: GameAction
): GameState {
  switch (action.type) {
    case RESET_GAME:
      return initialGameState;
    case COLLECT_ITEM:
      const collectedItem = state.board[action.payload.index];
      return {
        ...state,
        board: [
          ...state.board.slice(0, action.payload.index),
          ...state.board.slice(action.payload.index + 1),
        ],
        collected: collectedItem
          ? [...state.collected, collectedItem]
          : state.collected,
      };
    default:
      return state;
  }
}
