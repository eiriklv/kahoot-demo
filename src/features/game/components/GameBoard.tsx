import React from "react";

import { BoardItem } from "../duck";
import { GameBoardItem, GameBoardWrapper } from "../components/primitives";

interface GameBoardProps {
  gameBoard: BoardItem[];
  onItemClick: (index: number) => void;
}

export const GameBoard: React.FC<GameBoardProps> = ({
  gameBoard,
  onItemClick,
}) => {
  const boardItemElements = gameBoard.map((boardItem, index) => (
    <GameBoardItem
      key={`${index}-${boardItem}`}
      onClick={() => onItemClick(index)}
    >
      {boardItem}
    </GameBoardItem>
  ));

  return <GameBoardWrapper>{boardItemElements}</GameBoardWrapper>;
};
