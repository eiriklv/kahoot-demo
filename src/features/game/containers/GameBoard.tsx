import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { getGameBoard, collectItem } from "../duck";

import {
  GameBoardItem,
  GameBoardContainer,
  GameBoardHeader,
  GameBoardItemsContainer,
  Heading,
} from "../components/primitives";

export const GameBoard: React.FC = (props) => {
  const gameBoard = useSelector(getGameBoard);
  const dispatch = useDispatch();

  const handleItemClick = (index: number) => {
    dispatch(collectItem(index));
  };

  const gameItemElements = gameBoard.map((boardItem, index) => (
    <GameBoardItem onClick={() => handleItemClick(index)}>
      {boardItem}
    </GameBoardItem>
  ));

  return (
    <GameBoardContainer>
      <GameBoardHeader>
        <Heading>Kahoot! Points</Heading>
      </GameBoardHeader>
      <GameBoardItemsContainer>{gameItemElements}</GameBoardItemsContainer>
    </GameBoardContainer>
  );
};
