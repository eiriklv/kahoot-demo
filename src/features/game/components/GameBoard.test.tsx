import React from "react";
import renderer from "react-test-renderer";
import { GameBoard } from "./GameBoard";
import { BoardItem } from "../duck";

describe("GameBoard", () => {
  test("renders correctly when empty", () => {
    const gameBoard: BoardItem[] = [];

    const onItemClick = () => {};

    const tree = renderer
      .create(<GameBoard gameBoard={gameBoard} onItemClick={onItemClick} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("renders correctly when filled", () => {
    const gameBoard: BoardItem[] = ["A", "B", "C", "D", "A", "B", "C"];
    const onItemClick = () => {};

    const tree = renderer
      .create(<GameBoard gameBoard={gameBoard} onItemClick={onItemClick} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
