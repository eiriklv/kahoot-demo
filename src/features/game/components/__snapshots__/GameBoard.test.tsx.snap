// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard renders correctly when empty 1`] = `
<div
  className="sc-eCApGN lgxGRR"
/>
`;

exports[`GameBoard renders correctly when filled 1`] = `
<div
  className="sc-eCApGN lgxGRR"
>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    A
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    B
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    C
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    D
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    A
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    B
  </div>
  <div
    className="sc-jSFkmK jEMJdT"
    onClick={[Function]}
  >
    C
  </div>
</div>
`;
